### ISSUE TITLE
Ensemble communication improvements: avoid tensor copies

### ISSUE THREAD

**Issue by SlipknotTN** [NVIDIA/tensorrt-inference-server#343]
I have tried ensemble to call more models in a single request, mixing custom modules and DNN models.

It works, but I still pay lot of time to transfer data (input read or output write) between inner modules.
Is there some solution to hide this transfer time between models? For example this issue about IPC in the roadmap? #25

**Comment by deadeyegoodwin** [comment]
Right now all tensor communication between ensemble models happens via copies through system (CPU) memory. We have a couple of improvements on our roadmap but no schedule yet:
1. When a tensor is created on GPU and then consumed on GPU, keep it in GPU memory instead of copying through system memory.
2. Allow the memory (CPU or GPU) where a tensor is produced to be read directly by the consumer(s) of that tensor. This would eliminate any copy (except when CPU->GPU or GPU->CPU is required).

I'll keep this open and mark as enhancement so other can see and comment. I'm also changing the title somewhat since these improvements are not IPC related.

**Comment by SlipknotTN** [comment]
> Right now all tensor communication between ensemble models happens via copies through system (CPU) memory.

Is it true also for custom cpp models? No network interface is involved? Because I noticed that most of the time is spent reading from payload and writing to output buffer. I need to transfer about 70 - 100 MB between models inside the ensemble and it takes too much time to match "realtime" requirements.

**Comment by deadeyegoodwin** [comment]
Within TRTIS (this includes the communication between models in an ensemble) there is no network-based communication. It all happens by memory copy (either within CPU memory or between GPU and CPU memory). We are going to work on "zero-copy" ensemble communication where possible in the future but for now we don't have it. If you are communicating in your ensemble between 2 custom backends you can fix it yourself for now by sending a pointer to the data instead of the actual data (and managing the lifetime of that data appropriately).

**Comment by GuanLuo** [comment]
With #638, the ensemble will keep tensors on GPU depending on whether the framework backend supports GPU input / ouput. This enhancement will be in 19.09, and it is also in master so that you can try it now by building from source.

Currently PyTorch, TensorRT and Custom backends are aware of GPU I/O, and thus they can exploit this enhancement. For custom backends, you will need to use the new set of APIs (version 2) to be able to work with GPU memory for I/O.

**Comment by mrjackbo** [comment]
Great work! Is the passing of tensors on GPU as implemented in #638 a zero-copy mechanism?

**Comment by deadeyegoodwin** [comment]
No. The ensemble is still required to do a GPU copy from the output buffer of the model into a "holding" buffer that will provide the input for the next model in the ensemble. This is necessary because the output buffer must be freed so that it can be used for another request. The copy is on the device and so will be much faster than a device-to-host copy.

This current process maximizes throughput in that it frees the output buffer as soon as possible so that the model can be reused for another request. We are aware that there are use cases that would prefer reduced throughput in exchange for decrease latency that a a zero-copy would provide... but currently the ensemble scheduler cannot operate in that mode.

**Comment by solarflarefx** [comment]
I would like to better understand this context.

Is this thread referring to the idea of working in a pipeline architecture where all intermediary data resides in the GPU?  In this context the input to the neural network would already be in the GPU, so it would be ideal to pass this directly to the model into a model (which itself resides in the GPU), and whose output also stays in the GPU.  Is it possible to do this in the current framework?  If not, is it planned for the future?

**Comment by deadeyegoodwin** [comment]
@solarflarefx what you describe is possible using the TRTIS shared memory REST/GRPC API. This API lets you provide input tensors to TRTIS that are already in GPU memory and allows TRTIS to put the outputs directly into GPU buffers that you provide. If you wanted to link directly to the TRTIS shared library (libtrtserver.so) you can also do the same thing with the TRTSERVER shared-library API. See:

https://docs.nvidia.com/deeplearning/sdk/tensorrt-inference-server-guide/docs/client.html#cuda-shared-memory
https://docs.nvidia.com/deeplearning/sdk/tensorrt-inference-server-guide/docs/library_api.html
